### DISCUSSION TITLE
Milvus ServiceMonitor does not show up as Prometheus Targets and Grafana Panel displays "No Data"

### DISCUSSION THREAD

**Discussion by ganderaj** [milvus-io/milvus#26967] [Q&A and General discussion] [answered]
I am an issue with my Milvus deployment deployed via milvus-helm helm chart. I had the Service Monitor enabled and Prometheus / Grafana stack installed per Milvus documentation.

I have confirmed that Targets under Status bar of Prometheus dashboard are all up and that Service Monitor is installed. I also confirmed that Prometheus is capturing data points by inspecting Prometheus "TSBD Status" and Grafana Default dashboards.

However, Targets available are all Promethus default targets and the Milvus ServiceMonitor does not show up and Grafana Dashboard published by Milvus is displaying "No Data".

Can you please assist me here ?

- [serviceMonitor-milvus-yml.txt](https://github.com/milvus-io/milvus/files/12571153/serviceMonitor-milvus-yml.txt)
- [Prometheus Targets](https://user-images.githubusercontent.com/41026744/266418187-585db7eb-e606-4465-b744-012656eb9297.jpeg)
- [Prometheus TSBD Status](https://user-images.githubusercontent.com/41026744/266418191-85b2fca5-3c34-45de-a5a6-c563407e069d.jpeg)
- [Milvus Grafana Dashboard](https://user-images.githubusercontent.com/41026744/266418458-652e579c-deb1-4f8f-a4ad-747a7cccdc9c.jpeg)

Please note that despite having a Milvus installation using milvus-helm chart with the name "milvus", the "Instance" field on the Grafana Dashboard shows no available releases.
- [Helm Releases](https://user-images.githubusercontent.com/41026744/266429635-ce73fc85-7c43-41c2-b74a-035d3dd3f1e9.png)
- [Grafana Milvus Instances](https://user-images.githubusercontent.com/41026744/266429873-22546df5-0281-4ba0-a73d-839a21978c38.png)

I have tried creating a similar `Kind: serviceMonitor` [Service Monitors](https://user-images.githubusercontent.com/41026744/266445667-62d441c9-733a-44bb-8e0f-0a04f122d780.png) under `monitoring` namespace just to rule out if this has anything to do with namespace. But to no avail.

Could you please guide what could potentially cause `serviceMonitor:milvus/milvus` to not show up on Prometheus/Targets ?

**Comment by ganderaj** [comment]
@LoveEachDay @zwd1208 

Could you please help me here ?

**Comment by LoveEachDay** [comment]
@ganderaj From the attached prometheus targets page, we can see that prometheus operator doesn't add the milvus servicemonitor into prometheus. Could you provide the log of `prometheus-operator`?

**Reply by ganderaj** [reply]
`prometheus-operator` pod has two containers: `prometheus-operator` and `kube-rbac-proxy`. Logs have been pushed to CloudWatch using Fluent Bit. Attached are the logs I retrieved.

[prometheus-operator_kube-rbac-proxy.csv](https://github.com/milvus-io/milvus/files/12578441/prometheus-operator_kube-rbac-proxy.csv)
[prometheus-operator_prometheus-operator.csv](https://github.com/milvus-io/milvus/files/12578442/prometheus-operator_prometheus-operator.csv)

**Comment by ganderaj** [comment]
@LoveEachDay 

It appears no ServiceMonitors besides the default SMs installed by Prometheus are actually working. Is there any configuration we need to keep in mind while creating a SM manually ? Does the change of namespace effects scope of Prometheus to pull data ?

**Reply by ganderaj** [reply]
@LoveEachDay 

Our PRD Milvus deployment (helm release: _"milvus"_) is under _"milvus"_ namespace and its corresponding Service Monitor does not appear to show up as a Prometheus Target. To rule out namespace issue, we have created a test milvus deployment (helm release: _"milvus-test"_) under _"default"_ namespace and its corresponding namespace is working fine even a manually created SM pointing to the same under _"monitoring"_ namespace also works fine.

This basically tells me that Prometheus by default does not support custom namespaces besides the namespaces _"default"_ and _"monitoring"_, is it ? Is there any Prometheus configuration which requires an update ?

<img width="681" alt="image" src="https://github.com/milvus-io/milvus/assets/41026744/c70d1729-8cb8-455b-b9de-b6891897f050">

**Comment by mydreambei-ai** [comment, marked as answer]
@ganderaj  i have the same issue.  I found that the main reason for this issue is the lack of a "RoleBinding". If you milvus under the milvus namepsace, you can create role, rolebinding  for example:
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: lmy
  namespace: milvus-gpu
rules:
- apiGroups:
  - ""
  resources:
  - services
  - endpoints
  - pods
  verbs:
  - get
  - list
------
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: lmy
  namespace: milvus-gpu
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: lmy
subjects:
- kind: ServiceAccount
  name: prometheus-k8s (note: must be promethus serviceaccount)
  namespace: monitoring (note: must be promethus namepsace)


when you have create it, then you can kubectl get pods --n <milvus-namepace> --as=system:serviceaccount:<prometheus namespace>:<prometheus-serviceaccount>

**Reply by ganderaj** [reply]
This is awesome !
I am now able to find my SM as a Target post creating corresponding Role and RoleBinding. Many thanks !! 🫂

**Comment by ganderaj** [comment]
Raised a bug; can be addressed with a document update and/or a Helm Chart update.
https://github.com/milvus-io/milvus/issues/27088
